File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/internals/telemetry/logging/Logging.java
package org.firstinspires.ftc.teamcode.internals.telemetry.logging;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.telemetry.MultipleTelemetry;
import org.firstinspires.ftc.robotcore.external.Telemetry;

public class Logging {

    public static MultipleTelemetry TELEMETRY;

    public static void setTelemetry(Telemetry driver) {
        TELEMETRY = new MultipleTelemetry(driver, FtcDashboard.getInstance().getTelemetry());
    }

    public static void update() {
        TELEMETRY.update();
    }

    public static void clear() {
        TELEMETRY.clear();
    }

    public static void logData(String key, Object value) {
        TELEMETRY.addData(key, value);
    }

    public static void logText(String msg) {
        TELEMETRY.addLine(msg);
    }

    public static void log(String msg) {
        logText(msg);
    }

    public static void log(String key, Object value) {
        logData(key, value);
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/internals/telemetry/Logging.kt
package org.firstinspires.ftc.teamcode.internals.telemetry

import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.telemetry.MultipleTelemetry
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.vuforia.Vuforia.isInitialized
import org.firstinspires.ftc.robotcore.external.BlocksOpModeCompanion.telemetry
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.teamcode.internals.hardware.HardwareGetter.Companion.isEmulated

class Logging {
    companion object {

        @JvmStatic
        lateinit var driverTelemetry: Telemetry
        var dashboardPacket = TelemetryPacket()
        private lateinit var _telemetry: Telemetry;

        @JvmStatic
        val telemetry: Telemetry
            get() {
                if(!::driverTelemetry.isInitialized) {
                    throw UninitializedPropertyAccessException("driverTelemetry must be initialized to access telemetry!")
                }
                if(!::_telemetry.isInitialized) {
                    _telemetry = MultipleTelemetry(driverTelemetry, FtcDashboard.getInstance().telemetry)
                }
                return telemetry
            }

        @JvmStatic
        fun log(text: String) {
            logText(text)
        }

        @JvmStatic
        fun log(key: String, value: Any) {
            logData(key, value)
        }

        @JvmStatic
        fun logText(text: String) {
            if (isEmulated) {
                println(text)
            } else {
                driverTelemetry.addLine(text)
                dashboardPacket.addLine(text)
            }
        }

        @JvmStatic
        fun logData(key: String, value: Any) {
            if (isEmulated) {
                println("$key: $value")
            } else {
                driverTelemetry.addData(key, value)
                dashboardPacket.addLine(" ")
                dashboardPacket.put(key, value)
                dashboardPacket.put("$key (key)", "$key:")
            }
        }

        @JvmStatic
        fun updateLog() {
            if (!isEmulated) {
                driverTelemetry.update()
                FtcDashboard.getInstance().sendTelemetryPacket(dashboardPacket)
                dashboardPacket = TelemetryPacket()
            }
        }

        @JvmStatic
        fun clear() {
            if (!isEmulated) {
                driverTelemetry.clear()
                FtcDashboard.getInstance().clearTelemetry()
                dashboardPacket = TelemetryPacket()
            }
        }
    }
}